### ISSUE TITLE
always_specify_types should respect @optionalTypeArgs w.r.t static extensions

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#1706]
always_specify types allows _unspecified_ types when a class is  annotated with `@optionalTypeArgs`. This should be updated for extension methods. Some of this work might just be adding tests for verification.

* [ ] If `Baz<T>` is an extension on a class, `C<T>`, which has been marked `@optionalTypeArgs`, then `Baz("ok")` is ok;
  *  [ ] otherwise a Lint should be reported.
* [ ] The extension, `extension Baz on C`, where `C` is generic (e.g. `C<T>`), which has been marked `@optionalTypeArgs` is ok;
  * [ ] otherwise a Lint should be reported; `Baz` should have been declared as `extension Baz<T> on C<T>` or `extension Baz on C<dynamic>`, etc.
* [ ] The extension, `extension on C`, where `C` is generic (e.g. `C<T>`) and has been marked `@optionalTypeArgs` is ok;
  * [ ] otherwise  a Lint should be reported; the extension should have been declared as `extension on C<dynamic>`, etc., or should  have been named and generic, as above.

(This has been spun off of https://github.com/dart-lang/sdk/issues/38105)

**Comment by eernstg** [comment]
The part about `Baz("ok")` is a little bit confusing. If we have

```dart
class C<T> {...}
extension Baz<T> on C<T> {
  void baz() {}
}
```

then `Baz(C<int>()).baz()` would be an example of an explicit extension method invocation where some type arguments could be given, but were omitted. So that's a situation where `always_specify_types` could flag the missing type arguments (to `Baz`), and it could then silently accept the same thing when `Baz` has an `@optionalTypeArgs` in its metadata.

But isn't that an overly strict thing to do? I'd expect extension method invocations to be implicit almost all the time, and I would certainly expect a lint that forces developers to change `myC.baz()` to `Baz<int>(myC).baz()` to be too strict in practice, because extension method invocations _should_ be implicit whenever possible.

Another point in the design space would be to accept `myC.baz()` and `Baz<int>(myC).baz()`, but reject `Baz(myC).baz()`, but it seems odd to be really strict with the explicit invocation if there's no lint on `myC.baz()` (where exactly the same type arguments are omitted and inferred).

**Comment by srawlins** [comment]
This is an extremely strict lint rule. See the tests: https://github.com/dart-lang/linter/blob/master/test/rules/always_specify_types.dart

```dart
Map<String, String> map = {}; //LINT
```

But I have no skin in this game.

**Comment by AlexV525** [comment]
I'm facing the false positive issue with `@optionalTypeArgs` recently with the CodeFactor, and they've suggested this issue.

The problem is https://www.codefactor.io/repository/github/fluttercandies/flutter_wechat_assets_picker/issues, where the `ExamplePageMixin` is annotated with `@optionalTypeArgs`.

**Comment by AlexV525** [comment]
Sorry I missed some info. The annotation didn't fail locally. However the CodeFactor reports that I need to import `meta` explicitly. And after the import, the analyzer complains that it's already included in the import of `flutter/material.dart`.

I'll paste the conversation later.

**Comment by AlexV525** [comment]
So to solve the issue in the CodeFactor, I have to import `meta/meta.dart`. However:

![image](https://user-images.githubusercontent.com/15884415/161183251-fed7a4fa-71b2-4a09-b6f8-5be215eb7f4d.png)

By the time I thought it was included, they told me that the static analysis was not able to recognize the import:

![image](https://user-images.githubusercontent.com/15884415/161183410-384a1da7-de4b-4520-8ed1-5d2f3de16286.png)

Not sure if the above contents would help.

**Comment by pq** [comment]
> So to solve the issue in the CodeFactor, I have to import meta/meta.dart.

Strange.  On the surface this seems like a CodeFactor issue and might have something to do with how they're passing resolved packages to (what I assume) is a wrappered linter that they're consuming as a library (and not through the analyzer).  If you open an issue w/ them could you cc me and I'm happy to chime in?

Thanks!

**Comment by AlexV525** [comment]
> If you open an issue w/ them could you cc me and I'm happy to chime in?

I can open a new one instead. Which address should I cc you?
